README.md
qnfs [![Build Status](https://secure.travis-ci.org/fengmk2/qnfs.png)](http://travis-ci.org/fengmk2/qnfs) [![Coverage Status](https://coveralls.io/repos/fengmk2/qnfs/badge.png)](https://coveralls.io/r/fengmk2/qnfs) [![Build Status](https://drone.io/github.com/fengmk2/qnfs/status.png)](https://drone.io/github.com/fengmk2/qnfs/latest)
=======

![logo](https://raw.github.com/fengmk2/qnfs/master/logo.png)

Network Node.js [fs](http://nodejs.org/api/fs.html) module base on [qn](https://github.com/fengmk2/qn)

## Install

```bash
$ npm install qnfs
```

## Usage

```js
var fs = require('qnfs');
fs.config({
  accessKey: 'your access key',
  secretKey: 'your secret key',
  bucket: 'your bucket name',
  // timeout: 3600000, // default rpc timeout: one hour, optional
});

fs.stat('/qnfs/README.md', function (err, stat) {
  console.log(stat);
  // size: 527,
  // atime: Mon, 10 Oct 2011 23:24:11 GMT,
  // mtime: Mon, 10 Oct 2011 23:24:11 GMT,
  // ctime: Mon, 10 Oct 2011 23:24:11 GMT
});
```

## Documents

### `fs.stat(path, callback)`

### `fs.writeFile(filename, data, [options], callback)`

### `fs.readFile(filename, [options], callback)`

### `fs.appendFile(filename, data, [options], callback)`

### `fs.rename(oldPath, newPath, callback)`

### `fs.truncate(path, len, callback)`

### `fs.unlink(path, callback)`

### `fs.mkdir(path, [mode], callback)`

### `fs.readdir(path, callback)`

### `fs.rmdir(path, callback)`

### `fs.utimes(path, atime, mtime, callback)`

### `fs.exists(path, callback)`

### `fs.createReadStream(path, [options])`

### `fs.createWriteStream(path, [options])`

## TODO

* [ ] All async APIs on [Node.js fs](http://nodejs.org/api/fs.html)

## Authors

```bash
$ git summary 

 project  : qnfs
 repo age : 4 days
 active   : 3 days
 commits  : 13
 files    : 15
 authors  : 
    13  fengmk2                 100.0%
```

## License 

(The MIT License)

Copyright (c) 2013 fengmk2 &lt;fengmk2@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.0.1 / 2013-09-09 
==================

  * appendFile()
  * rename()
  * readdir()
  * unlink()
  * readFile()
  * add mkdir()
  * support exists()
  * update doc
  * add logo
  * add stat() and writeFile()
  * init


test/qnfs.test.js
/*!
 * qnfs - test/qnfs.test.js
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var path = require('path');
var fs = require('fs');
var should = require('should');
var pedding = require('pedding');
var qnfs = require('../');

var fixtures = path.join(__dirname, 'fixtures');
var CI_ENV = (process.env.TRAVIS ? 'TRAVIS' : process.env.CI_ENV) + '-' + process.version;
var fooFilePath = '/qnfs/test/fixtures/foo.txt.' + CI_ENV + '.txt';
var fooContent = fs.readFileSync(path.join(fixtures, 'foo.txt'));

describe('qnfs.test.js', function () {
  qnfs.config(require('./config.json'));

  before(function (done) {
    qnfs.writeFile(fooFilePath, fooContent, done);
  });

  describe('exists()', function () {
    it('should return true when file exists', function (done) {
      qnfs.exists(fooFilePath, function (exists) {
        exists.should.equal(true);
        done();
      });
    });

    it('should return false when file not exists', function (done) {
      qnfs.exists('/qnfs/test/fixtures/foo_not_exists.txt', function (exists) {
        exists.should.equal(false);
        done();
      });
    });

    it('should return true when dir exists', function (done) {
      qnfs.exists('/qnfs/test/fixtures/', function (exists) {
        exists.should.equal(true);
        done();
      });
    });

    it('should return false when dir not exists', function (done) {
      qnfs.exists('/qnfs/test/fixtures_not_exists/', function (exists) {
        exists.should.equal(false);
        done();
      });
    });
  });

  describe('rename()', function () {
    beforeEach(function (done) {
      done = pedding(2, done);
      qnfs.writeFile(fooFilePath + '.rename', fooContent, done);
      qnfs.unlink(fooFilePath + '.newname', function () {
        done();
      });
    });

    it('should rename a exists file', function (done) {
      qnfs.rename(fooFilePath + '.rename', fooFilePath + '.newname', function (err) {
        should.not.exist(err);
        qnfs.exists(fooFilePath + '.newname', function (exists) {
          exists.should.equal(true);
          done();
        });
      });
    });

    it('should rename same file work', function (done) {
      qnfs.rename(fooFilePath + '.rename', fooFilePath + '.rename', function (err) {
        should.not.exist(err);
        done();
      });
    });

    it('should rename not exists file return err', function (done) {
      qnfs.rename(fooFilePath + '.not-exists', fooFilePath + '.newname', function (err) {
        should.exist(err);
        err.name.should.equal('QiniuFileNotExistsError');
        err.message.should.equal('no such file or directory');
        // err.message.should.equal("ENOENT, rename '" + fooFilePath + '.not-exists' + "'");
        // console.log(err)
        done();
      });
    });
  });

  describe('mkdir()', function () {
    var dirpath = '/qnfs/test/mkdir' + CI_ENV;
    beforeEach(function (done) {
      qnfs.rmdir(dirpath, function (err) {
        done();
      });
    });

    it('should mkdir success', function (done) {
      qnfs.mkdir(dirpath, function (err, r) {
        should.not.exist(err);
        qnfs.readFile(r.key, function (err, data) {
          should.not.exists(err);
          var meta = JSON.parse(data);
          meta.should.have.keys('mode', 'path', 'ctime');
          meta.path.should.equal(dirpath + '/');
          meta.ctime.should.be.a('number');
          should.not.exist(meta.mode);
          done();
        });
      });
    });

    it('should mkdir mode:0666 success', function (done) {
      qnfs.mkdir(dirpath, '0666', function (err, r) {
        should.not.exist(err);
        qnfs.readFile(r.key, function (err, data) {
          should.not.exists(err);
          var meta = JSON.parse(data);
          meta.should.have.keys('mode', 'path', 'ctime');
          meta.path.should.equal(dirpath + '/');
          meta.ctime.should.be.a('number');
          meta.mode.should.equal('0666');
          done();
        });
      });
    });
  });

  describe('readdir()', function () {
    before(function (done) {
      done = pedding(7, done);
      qnfs.writeFile('/qnfs/test/readdir/foo.txt', fooContent, done);
      qnfs.writeFile('/qnfs/test/readdir/.foobar.txt', fooContent, done);
      qnfs.writeFile('/qnfs/test/readdir/subdir2/foo2.txt', fooContent, done);
      qnfs.writeFile('/qnfs/test/readdir/subdir2/foo3.txt', fooContent, done);
      qnfs.mkdir('/qnfs/test/readdir/subdir1', function () {
        done();
      });
      qnfs.mkdir('/qnfs/test/readdir/subdir1/subsubdir1', function () {
        done();
      });
      qnfs.mkdir('/qnfs/test/readdir/subdir2/subsubdir1/subdir3', function () {
        done();
      });
    });

    it('should return current dir all file and subdir names', function (done) {
      qnfs.readdir('/qnfs/test/readdir', function (err, names) {
        should.not.exist(err);
        should.exist(names);
        names.should.length(4);
        names.should.eql(['.foobar.txt', 'foo.txt', 'subdir1', 'subdir2']);
        done();
      });
    });

    it('should return []', function (done) {
      qnfs.readdir('/qnfs/test/readdir/subdir2/subsubdir1/subdir3', function (err, names) {
        should.not.exist(err);
        names.should.eql([]);
        done();
      });
    });

    it('should return error when dir not exists', function (done) {
      qnfs.readdir('/qnfs/test/readdir/not_exists', function (err) {
        should.exist(err);
        err.name.should.equal('Error');
        err.message.should.equal("ENOENT, readdir '/qnfs/test/readdir/not_exists/'");
        err.errno.should.equal(34);
        err.code.should.equal('ENOENT');
        done();
      });
    });
  });

  describe('writeFile()', function () {
    it('should return err when write a dir', function (done) {
      qnfs.writeFile('/tmp/', 'ddd', function (err) {
        should.exist(err);
        done();
      });
    });
  });

  describe('appendFile()', function () {
    before(function (done) {
      done = pedding(3, done);
      qnfs.unlink(fooFilePath + '.appendFile', function () {
        qnfs.writeFile(fooFilePath + '.appendFile', fooContent, done);
      });
      qnfs.unlink(fooFilePath + '.appendFile.empty', function () {
        qnfs.writeFile(fooFilePath + '.appendFile.empty', '', done);
      });      
      qnfs.unlink(fooFilePath + '.appendFile.not_exists', function () {
        done();
      });
    });

    it('should append content to not exists file', function (done) {
      qnfs.appendFile(fooFilePath + '.appendFile.not_exists', 'new content', function (err) {
        should.not.exist(err);
        qnfs.readFile(fooFilePath + '.appendFile.not_exists', function (err, data) {
          should.not.exist(err);
          data.toString().should.equal('new content');
          done();
        });
      });
    });

    it('should append content to exists file', function (done) {
      qnfs.appendFile(fooFilePath + '.appendFile', 'new content', function (err) {
        should.not.exist(err);
        qnfs.readFile(fooFilePath + '.appendFile', function (err, data) {
          should.not.exist(err);
          data.toString().should.equal('This is a foo content.new content');
          done();
        });
      });
    });

    it('should append content to exists empty file', function (done) {
      qnfs.appendFile(fooFilePath + '.appendFile.empty', 'new content', function (err) {
        should.not.exist(err);
        qnfs.readFile(fooFilePath + '.appendFile.empty', function (err, data) {
          should.not.exist(err);
          data.toString().should.equal('new content');
          done();
        });
      });
    });

    it('should return err when append to dir', function (done) {
      qnfs.appendFile('/tmp/', 'ddddd', function (err) {
        should.exist(err);
        done();
      });
    });
  });

  describe('readFile()', function () {
    it('should return file content buffer', function (done) {
      qnfs.readFile(fooFilePath, function (err, content) {
        should.not.exist(err);
        should.ok(Buffer.isBuffer(content));
        content.toString().should.equal(fooContent.toString());
        done();
      });
    });

    it('should return file content string', function (done) {
      qnfs.readFile(fooFilePath, 'utf8', function (err, content) {
        should.not.exist(err);
        content.should.be.a('string');
        content.should.equal(fooContent.toString());
        done();
      });
    });

    it('should return error when file not exists', function (done) {
      qnfs.readFile(fooFilePath + '_not_exists', 'utf8', function (err, content) {
        should.exist(err);
        err.name.should.equal('QiniuNotFoundError');
        err.message.should.equal("ENOENT, open '" + fooFilePath + '_not_exists' + "'");
        err.errno.should.equal(34);
        err.code.should.equal('ENOENT');
        done();
      });
    });

    it('should return error when filename is dir', function (done) {
      qnfs.readFile('/tmp/', 'utf8', function (err, content) {
        should.exist(err);
        err.name.should.equal('Error');
        err.message.should.equal("EISDIR, read");
        err.errno.should.equal(28);
        err.code.should.equal('EISDIR');
        done();
      });
    });
  });

  describe('unlink()', function () {
    before(function (done) {
      qnfs.writeFile(fooFilePath + '.unlink', fooContent, done);
    });

    it('should unlink a file', function (done) {
      qnfs.exists(fooFilePath + '.unlink', function (exists) {
        should.ok(exists);
        qnfs.unlink(fooFilePath + '.unlink', function (err) {
          should.not.exists(err);
          qnfs.exists(fooFilePath + '.unlink', function (exists) {
            exists.should.equal(false);
            done();
          });
        });
      });
    });

    it('should unlink a not exists return error', function (done) {
      qnfs.unlink(fooFilePath + '.unlink_not_exists', function (err) {
        should.exists(err);
        err.name.should.equal('QiniuFileNotExistsError');
        err.message.should.equal("ENOENT, unlink '" + fooFilePath + '.unlink_not_exists' + "'");
        err.errno = 34;
        err.code = 'ENOENT';
        done();
      });
    });

    it('should unlink a dir return error', function (done) {
      qnfs.unlink('/tmp/', function (err) {
        should.exists(err);
        err.name.should.equal('Error');
        err.message.should.equal("EACCES, unlink '/tmp/'");
        err.errno = 3;
        err.code = 'EACCES';
        err.path.should.equal('/tmp/');
        done();
      });
    });
  });

  describe('stat()', function () {
    it('should get a file stat', function (done) {
      qnfs.stat(fooFilePath, function (err, stat) {
        should.not.exist(err);
        should.exist(stat);
        stat.should.have.keys('size', 'atime', 'mtime', 'ctime');
        stat.size.should.equal(fs.statSync(path.join(fixtures, 'foo.txt')).size);
        done();
      });
    });

    it('should return error when file not exists', function (done) {
      qnfs.stat('/qnfs/test/fixtures/foo.not.exists.txt', function (err, stat) {
        should.exist(err);
        err.name.should.equal('QiniuFileNotExistsError');
        err.message.should.equal('no such file or directory');
        should.not.exist(stat);
        done();
      });
    });
  });
});


lib/stat.js
/*!
 * qnfs - lib/stat.js
 *
 * Impl fs.Stat all APIs: http://nodejs.org/docs/latest/api/fs.html#fs_class_fs_stats
 *
 * stats.isFile()
 * stats.isDirectory()
 * stats.isBlockDevice()
 * stats.isCharacterDevice()
 * stats.isSymbolicLink() (only valid with fs.lstat())
 * stats.isFIFO()
 * stats.isSocket()
 * 
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

function Stat(info) {
  var ctime = new Date(info.putTime / 10000);
  this.size = info.fsize;
  this.atime = new Date();
  this.mtime = ctime;
  this.ctime = ctime;
}

Stat.prototype.isFile = function () {
  return true;
};

Stat.prototype.isDirectory =
Stat.prototype.isBlockDevice =
Stat.prototype.isCharacterDevice =
Stat.prototype.isSymbolicLink =
Stat.prototype.isFIFO =
Stat.prototype.isSocket = function () {
  return false;
};

module.exports = Stat;


lib/qnfs.js
/*!
 * qnfs - lib/qnfs.js
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var debug = require('debug')('qnfs');
var pathlib = require('path');
var qn = require('qn');
var Stat = require('./stat');
var utils = require('./utils');

var client;

exports.config = function (options) {
  client = exports.client = qn.create(options);
};

/**
 * Stat
 */

exports.stat = function (filename, callback) {
  client.stat(filename, function (err, info) {
    if (err) {
      return callback(err);
    }
    callback(null, new Stat(info));
  });
};

/**
 * Exists
 */

/**
 * Test whether or not the given dir path exists by checking with the file system. 
 * Then call the callback argument with either true or false.
 * 
 * @param {String} path dir path, must end with "/", e.g.: "/foo/", "/foo/bar/"
 * @param {Function(exists)} callback
 *  - {Boolean} exists
 */
exports.existsDir = function (path, callback) {
  if (!utils.isDirPath(path)) {
    return callback(false);
  }
  client.list({prefix: path, limit: 1}, function (err, result) {
    callback(!err && result && result.items && result.items.length > 0);
  });
};

/**
 * Test whether or not the given path exists by checking with the file system. 
 * Then call the callback argument with either true or false.
 * 
 * @param {String} path if you want to test dir path, must end with "/", e.g.: "/foo/", "/foo/bar/"
 * @param {Function(exists)} callback
 *  - {Boolean} exists
 */
exports.exists = function (path, callback) {
  if (utils.isDirPath(path)) {
    return exports.existsDir(path, callback);
  }

  exports.stat(path, function (err, stat) {
    callback(!err && !!stat);
  });
};

/**
 * Write
 */

exports.writeFile = function (filename, data, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = null;
  }

  // { [Error: EISDIR, open '/tmp'] errno: 28, code: 'EISDIR', path: '/tmp' }
  if (utils.isDirPath(filename)) {
    var err = new Error('EISDIR, open \'' + filename + '\'');
    err.errno = 28;
    err.code = 'EISDIR';
    err.path = filename;
    return callback(err);
  }

  exports.stat(filename, function (err, stat) {
    if (err && err.name !== 'QiniuFileNotExistsError') {
      return callback(err);
    }

    if (stat) {
      client.delete(filename, function (err) {
        if (err) {
          return callback(err);
        }

        client.upload(data, {key: filename}, callback);
      });
      return;
    }

    client.upload(data, {key: filename}, callback);
  });
};

/**
 * Append
 */

/**
 * Asynchronously append data to a file, creating the file if it not yet exists. data can be a string or a buffer.
 * @param {String} filename
 * @param {String|Buffer} data
 * @param {Object|String} [options]
 *  - {String|Null} encoding default = 'utf8'
 * @param {Function(err)} callback
 */
exports.appendFile = function (filename, data, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = null;
  }

  // { [Error: EISDIR, open '/tmp'] errno: 28, code: 'EISDIR', path: '/tmp' }
  if (utils.isDirPath(filename)) {
    var err = new Error('EISDIR, open \'' + filename + '\'');
    err.errno = 28;
    err.code = 'EISDIR';
    err.path = filename;
    return callback(err);
  }

  var encoding = options || 'utf8';
  if (options) {
    encoding = options.encoding;
  }
  if (typeof data === 'string') {
    data = new Buffer(data, encoding);
  }

  exports.readFile(filename, function (err, buf) {
    if (err) {
      if (err.code === 'ENOENT') {
        return exports.writeFile(filename, data, callback);
      } else {
        return callback(err);
      }
    }

    if (buf && buf.length > 0) {
      data = Buffer.concat([buf, data]);
    }
    // delete old file
    exports.unlink(filename, function (err) {
      if (err) {
        return callback(err);
      }

      exports.writeFile(filename, data, callback);
    });
  });
};

/**
 * Asynchronous truncate(2). No arguments other than a possible exception are given to the completion callback.
 * 
 * @param {String} filename
 * @param {Number} len
 * @param {Function(err)} callback
 */
exports.truncate = function (filename, len, callback) {
  // TODO
};

/**
 * Read
 */

/**
 * Asynchronously reads the entire contents of a file.
 * 
 * @param {String} filename
 * @param {Object} [options]
 *  - {String|Null} encoding default = null
 *  - {String} flag default = 'r'
 * @param {Function(err, data)} callback
 */
exports.readFile = function (filename, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = null;
  }

  var encoding = options;
  if (options && options.encoding) {
    encoding = options.encoding;
  }

  // read a dir
  // { [Error: EISDIR, read] errno: 28, code: 'EISDIR' }
  if (utils.isDirPath(filename)) {
    var err = new Error('EISDIR, read');
    err.errno = 28;
    err.code = 'EISDIR';
    return callback(err);
  }

  client.download(filename, function (err, data) {
    if (err) {
      // QiniuNotFoundError
      // { [Error: ENOENT, open 'sdfsdf'] errno: 34, code: 'ENOENT', path: 'sdfsdf' }
      if (err.name === 'QiniuNotFoundError') {
        err.message = 'ENOENT, open \'' + filename + '\'';
        err.code = 'ENOENT';
        err.errno = 34;
        err.path = filename;
      }
      return callback(err);
    }
    if (encoding) {
      data = data.toString(encoding);
    }
    callback(null, data);
  });
};


/**
 * Rename
 */

exports.rename = function (oldPath, newPath, callback) {
  debug('rename %s to %s', oldPath, newPath);

  // { [Error: ENOENT, rename 'foo'] errno: 34, code: 'ENOENT', path: 'foo' }
  client.move(oldPath, newPath, callback);
};

/**
 * Unlink
 */

exports.unlink = function (path, callback) {
  debug('unlink %s', path);
  // not exists: { [Error: ENOENT, unlink 'adawda'] errno: 34, code: 'ENOENT', path: 'adawda' }
  // unlink dir: { [Error: EACCES, unlink '/tmp'] errno: 3, code: 'EACCES', path: '/tmp' }
  if (utils.isDirPath(path)) {
    var err = new Error('EACCES, unlink \'' + path + '\'');
    err.errno = 3;
    err.code = 'EACCES';
    err.path = path;
    return callback(err);
  }
  client.delete(path, function (err) {
    if (err && (err.name === 'QiniuNotFoundError' || err.name === 'QiniuFileNotExistsError')) {
      err.message = 'ENOENT, unlink \'' + path + '\'';
      err.errno = 34;
      err.code = 'ENOENT';
      err.path = path;
    }
    callback(err);
  });
};

/**
 * Make dir
 */

exports.mkdir = function (path, mode, callback) {
  if (typeof mode === 'function') {
    callback = mode;
    mode = null;
  }

  path = utils.ensureDir(path);

  exports.existsDir(path, function (exists) {
    // errno: 47, code: 'EEXIST', path: 'test'
    if (exists) {
      var err = new Error("EEXIST, mkdir '" + path + "'");
      err.errno = 47;
      err.code = 'EEXIST';
      err.path = path;
      return callback(err);
    }

    var meta = {
      path: path,
      mode: mode,
      ctime: Date.now()
    };
    exports.writeFile(utils.dirMetaFile(path), JSON.stringify(meta), callback);
  });
};

/**
 * Remove dir
 */

exports.rmdir = function (path, callback) {
  path = utils.ensureDir(path);

  client.list({prefix: path, limit: 2}, function (err, result) {
    if (err) {
      return callback(err);
    }

    // { [Error: ENOENT, rmdir 'res22'] errno: 34, code: 'ENOENT', path: 'res22' }
    if (!result || result.items.length === 0) {
      err = new Error("ENOENT, rmdir '" + path + "'");
      err.errno = 34;
      err.code = 'ENOENT';
      err.path = path;
      return callback(err);
    }

    var metafile = utils.dirMetaFile(path);

    // not empty
    if (result.items.length > 2 || ('/' + result.items[0].key) !== metafile) {
      // { [Error: ENOTEMPTY, rmdir 'res'] errno: 53, code: 'ENOTEMPTY', path: 'res' }
      err = new Error("ENOTEMPTY, rmdir '" + path + "'");
      err.errno = 53;
      err.code = 'ENOTEMPTY';
      err.path = path;
      return callback(err);
    }

    // delete metafile
    exports.unlink(metafile, callback);
  });
};

/**
 * Read dir
 */

/**
 * Asynchronous readdir(3). Reads the contents of a directory. 
 * The callback gets two arguments (err, files) where files is an array of the names of the files 
 * in the directory excluding '.' and '..'.
 * 
 * @param {String} path
 * @param {Function(err, names)} callback
 */
exports.readdir = function (path, callback) {
  path = utils.ensureDir(path);
  // { [Error: ENOENT, readdir 'sdfsdf'] errno: 34, code: 'ENOENT', path: 'sdfsdf' }
  // { [Error: ENOTDIR, readdir 'test/config.json'] errno: 27, code: 'ENOTDIR', path: 'test/config.json' }
  client.list({prefix: path, limit: 1000}, function (err, result) {
    // TODO: read all files
    if (err) {
      return callback(err);
    }

    var items = result.items || [];
    if (items.length === 0) {
      err = new Error("ENOENT, readdir '" + path + "'");
      err.errno = 34;
      err.code = 'ENOENT';
      err.path = path;
      return callback(err);
    }

    var names = {};
    for (var i = 0; i < items.length; i++) {
      var key = '/' + items[i].key.trim();
      if (utils.isDirMetaFile(key)) {
        key = pathlib.dirname(key);
      }
      var name = key.substring(path.length);
      var index = name.indexOf('/');
      if (index > 0) {
        name = name.substring(0, index);
      }
      if (!name) {
        continue;
      }

      names[name] = 1;
    }
    callback(null, Object.keys(names));
  });
};
